Mathlib/Computability/QueryComplexity/Defs.lean

```diff
@@ -0,0 +1,102 @@
+/-
+Copyright (c) 2025 Tomaz Mascarenhas. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Geoffrey Irving, Tomaz Mascarenhas
+-/
+import Mathlib.Algebra.Group.Pi.Basic
+import Mathlib.Data.Set.Basic
+
+/-!
+# Definition of a model of computation based on oracles.
+
+This module defines an abstraction of computation with oracles, enabling the analysis of upper
+bounds on the query complexity of algorithms. It also provides a Monad instance for it.
+
+Note that this is the deterministic version. In the future, the stochastic version will
+also be ported, which enables the possibility of proving lower bounds on the query complexity
+of algorithms. In order to facilitate this future integration, the design of the types
+in this module is based on the original stochastic version. Source:
+  https://github.com/girving/debate/blob/862fdb1cf55df0d541b802bdb1e672d724df6398/Comp/Oracle.lean
+
+## Main Definitions
+
+* Comp (ι : Type) {I : Type} (s : Set I) (α : Type) : Type
+* Comp.run : Comp ι s α → (I → Oracle ι) → α × (I → ℕ)
+-/
+
+open Classical
+open Set
+noncomputable section
+
+variable {ι I : Type}
+variable {s t : Set I}
+variable {α β γ : Type}
+
+namespace QueryComplexity
+
+/-- A deterministic oracle is a map from `α` to `Bool` -/
+def Oracle (α : Type) := α → Bool
+
+/-- A deterministic computation that make decisions by querying oracles. A computation is either a
+pure value or the identifier of an oracle (`o`) drawn from a predefined set (`s`), a value to
+be queried by the oracle (`i`) and two other computations, to be run depending on the answer
+of the oracle. -/
+inductive Comp (ι : Type) {I : Type} (s : Set I) (α : Type) : Type where
+  | pure' : α → Comp ι s α
+  | query' : (o : I) → o ∈ s → (i : ι) → Comp ι s α → Comp ι s α → Comp ι s α
```

Comment:
I got some significant simplifications out of replacing the two `Comp` arguments here with `Bool → Comp`. It practice it seems to reduce the number of cases needed in proofs.

See recent https://github.com/girving/debate for the new style.